File: typescript/src/gyro/signedFixedPoint.ts
class FixedPointError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'FixedPointError';
    }
}

export class SignedFixedPoint {
    public static readonly ONE = BigInt('1000000000000000000'); // 1e18
    public static readonly ONE_XP = BigInt(
        '100000000000000000000000000000000000000',
    ); // 1e38

    static add(a: bigint, b: bigint): bigint {
        const c = a + b;
        if (!(b >= 0n ? c >= a : c < a)) {
            throw new FixedPointError('AddOverflow');
        }
        return c;
    }

    static addMag(a: bigint, b: bigint): bigint {
        return a > 0n ? this.add(a, b) : this.sub(a, b);
    }

    static sub(a: bigint, b: bigint): bigint {
        const c = a - b;
        if (!(b <= 0n ? c >= a : c < a)) {
            throw new FixedPointError('SubOverflow');
        }
        return c;
    }

    static mulDownMag(a: bigint, b: bigint): bigint {
        const product = a * b;
        if (!(a === 0n || product / a === b)) {
            throw new FixedPointError('MulOverflow');
        }
        return product / this.ONE;
    }

    static mulDownMagU(a: bigint, b: bigint): bigint {
        return (a * b) / this.ONE;
    }

    static mulUpMag(a: bigint, b: bigint): bigint {
        const product = a * b;
        if (!(a === 0n || product / a === b)) {
            throw new FixedPointError('MulOverflow');
        }

        if (product > 0n) {
            return (product - 1n) / this.ONE + 1n;
        } else if (product < 0n) {
            return (product + 1n) / this.ONE - 1n;
        }
        return 0n;
    }

    static mulUpMagU(a: bigint, b: bigint): bigint {
        const product = a * b;
        if (product > 0n) {
            return (product - 1n) / this.ONE + 1n;
        } else if (product < 0n) {
            return (product + 1n) / this.ONE - 1n;
        }
        return 0n;
    }

    static divDownMag(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        if (a === 0n) {
            return 0n;
        }

        const aInflated = a * this.ONE;
        if (aInflated / a !== this.ONE) {
            throw new FixedPointError('DivInterval');
        }

        return aInflated / b;
    }

    static divDownMagU(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        return (a * this.ONE) / b;
    }

    static divUpMag(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        if (a === 0n) {
            return 0n;
        }

        let localA = a;
        let localB = b;
        if (b < 0n) {
            localB = -b;
            localA = -a;
        }

        const aInflated = localA * this.ONE;
        if (aInflated / localA !== this.ONE) {
            throw new FixedPointError('DivInterval');
        }

        if (aInflated > 0n) {
            return (aInflated - 1n) / localB + 1n;
        }
        return (aInflated + 1n) / localB - 1n;
    }

    static divUpMagU(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        if (a === 0n) {
            return 0n;
        }

        let localA = a;
        let localB = b;
        if (b < 0n) {
            localB = -b;
            localA = -a;
        }

        if (localA > 0n) {
            return (localA * this.ONE - 1n) / localB + 1n;
        }
        return (localA * this.ONE + 1n) / localB - 1n;
    }

    static mulXp(a: bigint, b: bigint): bigint {
        const product = a * b;
        if (!(a === 0n || product / a === b)) {
            throw new FixedPointError('MulOverflow');
        }
        return product / this.ONE_XP;
    }

    static mulXpU(a: bigint, b: bigint): bigint {
        return (a * b) / this.ONE_XP;
    }

    static divXp(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        if (a === 0n) {
            return 0n;
        }

        const aInflated = a * this.ONE_XP;
        if (aInflated / a !== this.ONE_XP) {
            throw new FixedPointError('DivInterval');
        }

        return aInflated / b;
    }

    static divXpU(a: bigint, b: bigint): bigint {
        if (b === 0n) {
            throw new FixedPointError('ZeroDivision');
        }
        return (a * this.ONE_XP) / b;
    }

    static mulDownXpToNp(a: bigint, b: bigint): bigint {
        const E19 = BigInt('10000000000000000000');
        const b1 = b / E19;
        const prod1 = a * b1;
        if (!(a === 0n || prod1 / a === b1)) {
            throw new FixedPointError('MulOverflow');
        }
        const b2 = b % E19;
        const prod2 = a * b2;
        if (!(a === 0n || prod2 / a === b2)) {
            throw new FixedPointError('MulOverflow');
        }
        return prod1 >= 0n && prod2 >= 0n
            ? (prod1 + prod2 / E19) / E19
            : (prod1 + prod2 / E19 + 1n) / E19 - 1n;
    }

    static mulDownXpToNpU(a: bigint, b: bigint): bigint {
        const E19 = BigInt('10000000000000000000');
        const b1 = b / E19;
        const b2 = b % E19;
        const prod1 = a * b1;
        const prod2 = a * b2;
        return prod1 >= 0n && prod2 >= 0n
            ? (prod1 + prod2 / E19) / E19
            : (prod1 + prod2 / E19 + 1n) / E19 - 1n;
    }

    static mulUpXpToNp(a: bigint, b: bigint): bigint {
        const E19 = BigInt('10000000000000000000');
        const b1 = b / E19;
        const prod1 = a * b1;
        if (!(a === 0n || prod1 / a === b1)) {
            throw new FixedPointError('MulOverflow');
        }
        const b2 = b % E19;
        const prod2 = a * b2;
        if (!(a === 0n || prod2 / a === b2)) {
            throw new FixedPointError('MulOverflow');
        }
        return prod1 <= 0n && prod2 <= 0n
            ? (prod1 + prod2 / E19) / E19
            : (prod1 + prod2 / E19 - 1n) / E19 + 1n;
    }

    static mulUpXpToNpU(a: bigint, b: bigint): bigint {
        const E19 = BigInt('10000000000000000000');
        const b1 = b / E19;
        const b2 = b % E19;
        const prod1 = a * b1;
        const prod2 = a * b2;
        return prod1 <= 0n && prod2 <= 0n
            ? (prod1 + prod2 / E19) / E19
            : (prod1 + prod2 / E19 - 1n) / E19 + 1n;
    }

    static complement(x: bigint): bigint {
        if (x >= this.ONE || x <= 0n) {
            return 0n;
        }
        return this.ONE - x;
    }
}

File: rust/src/pools/gyro/signed_fixed_point.rs
use alloy_primitives::{uint, I256};

pub const ONE: I256 = I256::from_raw(uint!(1000000000000000000_U256)); // 1e18
pub const ONE_E_19: I256 = I256::from_raw(uint!(10000000000000000000_U256)); // 1e19
pub const ONE_XP: I256 = I256::from_raw(uint!(100000000000000000000000000000000000000_U256)); // 1e38

#[derive(Debug)]
pub struct FixedPointError(pub &'static str);

pub fn mul_down_mag(a: &I256, b: &I256) -> I256 {
    (*a * *b) / ONE
}

pub fn mul_up_mag(a: &I256, b: &I256) -> I256 {
    let product = *a * *b;
    if product > I256::ZERO {
        (product - I256::ONE) / ONE + I256::ONE
    } else if product < I256::ZERO {
        (product + I256::ONE) / ONE - I256::ONE
    } else {
        I256::ZERO
    }
}

pub fn div_down_mag(a: &I256, b: &I256) -> I256 {
    if b.is_zero() {
        panic!("ZeroDivision");
    }
    if a.is_zero() {
        return I256::ZERO;
    }
    let a_inflated = *a * ONE;
    a_inflated / *b
}

pub fn div_up_mag(a: &I256, b: &I256) -> I256 {
    if b.is_zero() {
        panic!("ZeroDivision");
    }
    if a.is_zero() {
        return I256::ZERO;
    }
    let mut local_a = *a;
    let mut local_b = *b;
    if b < &I256::ZERO {
        local_b = -local_b;
        local_a = -local_a;
    }
    let a_inflated = local_a * ONE;
    if a_inflated > I256::ZERO {
        (a_inflated - I256::ONE) / local_b + I256::ONE
    } else {
        (a_inflated + I256::ONE) / local_b - I256::ONE
    }
}

// Signed versions of XP functions
pub fn mul_xp_u(a: &I256, b: &I256) -> I256 {
    (*a * *b) / ONE_XP
}

pub fn div_xp_u(a: &I256, b: &I256) -> I256 {
    if b.is_zero() {
        panic!("ZeroDivision");
    }
    (*a * ONE_XP) / *b
}

pub fn mul_down_xp_to_np(a: &I256, b: &I256) -> I256 {
    let b1 = *b / ONE_E_19;
    let b2 = *b % ONE_E_19;
    let prod1 = *a * b1;
    let prod2 = *a * b2;

    if prod1 >= I256::ZERO && prod2 >= I256::ZERO {
        let prod2_div_e19 = prod2 / ONE_E_19;
        (prod1 + prod2_div_e19) / ONE_E_19
    } else {
        let prod2_div_e19 = prod2 / ONE_E_19;
        (prod1 + prod2_div_e19 + I256::ONE) / ONE_E_19 - I256::ONE
    }
}

pub fn mul_up_xp_to_np(a: &I256, b: &I256) -> I256 {
    let b1 = *b / ONE_E_19;
    let b2 = *b % ONE_E_19;
    let prod1 = *a * b1;
    let prod2 = *a * b2;

    if prod1 <= I256::ZERO && prod2 <= I256::ZERO {
        let prod2_div_e19 = prod2 / ONE_E_19;
        (prod1 + prod2_div_e19) / ONE_E_19
    } else {
        let prod2_div_e19 = prod2 / ONE_E_19;
        (prod1 + prod2_div_e19 - I256::ONE) / ONE_E_19 + I256::ONE
    }
}
